url294e6c45f3_Ecdosis_NMergeNew_tgz-pJ8-TestJ8/src/edu/luc/nmerge/mvd/MVDXMLFile.java

```diff
@@ -0,0 +1,532 @@
+							llp.add( p );
+							orphans.put( pId, llp );
+						}
+						else
+							llp.add( p );
+					}
+					else
+						parent.addChild( p );
+				}
+				mvd.addPair( p );
+			}
+		}
+	}
+	/**
+	 * Build a BitSet of versions from a string where each version id 
+	 * is separated by a comma and runs separated by hyphens.
+	 * @param versions a string containing versions in a compact 
+	 * readable format
+	 * @return the finished BitSet corresponding to the inputs
+	 * @throws an Exception if the syntax is wrong
+	 */
+	private static BitSet buildVersion( String versions, 
+		boolean hint ) throws Exception
+	{
+		BitSet bs = new BitSet();
+		if ( hint )
+			bs.set( 0 );
+		String[] parts = versions.split(",");
+		for ( int i=0;i<parts.length;i++ )
+		{
+			if ( parts[i].indexOf('-') != -1 )
+			{
+				String[] subParts = parts[i].split("-");
+				if ( subParts.length > 2 )
+					throw new MVDException( "Invalid version run "
+						+parts[i] );
+				else
+				{
+					int start = Integer.parseInt( subParts[0] );
+					int end = Integer.parseInt( subParts[1] );
+					for ( int j=start;j<=end;j++ )
+						bs.set( j );
+				}
+			}
+			else
+				bs.set( Integer.parseInt(parts[i]) );
+		}
+		return bs;
+	}
+	/**
+	 * Write an MVD out as text. If there are angle brackets or 
+	 * ampersands in the text, these are automatically escaped 
+	 * as entities
+	 * @param mvd the source multi-version document
+	 * @param dst the destination MVD XML file
+	 * @param XMLEncoding the desired encoding of the output 
+	 * (not that of the source mvd)
+	 * @param srcEncoding the encoding of the data in the MVD
+	 * @param pretty if true make some attempt at tidying the output
+	 * @throws an exception if it couldn't write the file or find a 
+	 * serializer
+	 */
+	public static void externalise( MVD mvd, File dst, 
+		String XMLEncoding, String srcEncoding, boolean pretty ) 
+	throws Exception
+	{
+		DocumentBuilderFactory dbf 
+			= DocumentBuilderFactory.newInstance();
+		try 
+		{
+			DocumentBuilder db = dbf.newDocumentBuilder();
+			// create an instance of DOM
+			Document dom = db.newDocument();
+			Element rootElement = dom.createElement( "mvd" );
+			rootElement.setAttribute( "description", 
+				mvd.getDescription() );
+			rootElement.setAttribute( "encoding", 
+				mvd.getEncoding() );
+			dom.appendChild( rootElement );
+			writeGroups( dom, mvd, rootElement );
+			writeVersions( dom, mvd, rootElement );
+			writePairs( dom, mvd, rootElement, srcEncoding );
+			// print it out manually since DOM serialisation escapes 
+			// CRLF as &13;LF on the Mac and this behaviour can't be 
+			// turned off. We need a portable externalise routine that 
+			// works the same on all platforms
+			FileOutputStream fos = new FileOutputStream( dst );
+			XMLPretty.writeXMLDeclaration( fos, XMLEncoding );
+			XMLPretty.writeNode( dom.getFirstChild(), fos, 
+				XMLEncoding, pretty, 0 );
+			fos.close();
+		}
+		catch ( Exception ioe )
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)